runtime/router.go

```diff
@@ -128,8 +132,41 @@ func (endpoint *RouterEndpoint) HandleRequest(
 	urlValues := ParamsFromContext(r.Context())
 	req := NewServerHTTPRequest(w, r, urlValues, endpoint)
 	ctx := req.Context()
+
+	// setting up capture for endpoint
+	if endpoint.eventSampler(endpoint.EndpointName, endpoint.HandlerName) {
+		ctx = WithToCapture(ctx)
+		ctx = WithEventContainer(ctx, &EventContainer{})
+	}
+
 	endpoint.HandlerFn(ctx, req, req.res)
 	req.res.flush(ctx)
+
+	// retrieve the container to see if any events are generated?
+	if GetToCapture(ctx) {
+		ec := GetEventContainer(ctx)
+		var events []Event
+		if ec != nil {
+			events = append(events, ec.events...)
+		}
+
+		event := &HTTPIncomingEvent{
+			EndpointName: endpoint.EndpointName,
+			HandlerName:  endpoint.HandlerName,
+			HTTPCapture: HTTPCapture{
+				ReqURL:        r.URL.String(),
```

Comment:
i think we should clone req headers before handlerFn

Reply:
Headers weren't being mutated in the endpoint handler. However, it could change in the future. Capturing early is better. Have made the change.